eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs

```diff
@@ -1389,18 +1394,24 @@ impAnn msg m = do
 
 -- | Adds a source location and Doc to the log, which are only shown if the test fails
 logWithCallStack :: CallStack -> Doc AnsiStyle -> ImpTestM era ()
-logWithCallStack callStack e = impLogL %= (<> loc <> line <> indent 2 e <> line)
+logWithCallStack callStack entry = impLogL %= (<> stack <> line <> indent 2 entry <> line)
   where
-    formatSrcLoc srcLoc =
-      "[" <> srcLocModule srcLoc <> ":" <> show (srcLocStartLine srcLoc) <> "]"
-    loc =
-      case getCallStack callStack of
-        (_, srcLoc) : _ -> annotate (color Blue) . pretty $ formatSrcLoc srcLoc
-        _ -> mempty
-
--- | Adds a string to the log, which is only shown if the test fails
-logEntry :: HasCallStack => String -> ImpTestM era ()
-logEntry = logWithCallStack ?callStack . pretty
+    prettySrcLoc' SrcLoc {..} =
+      hcat
+        [ annotate (color c) d
+        | (c, d) <-
+            [ (Yellow, "[")
+            , (Blue, pretty srcLocModule)
+            , (Yellow, ":")
+            , (Magenta, pretty srcLocStartLine)
+            , (Yellow, "]")
+            ]
+        ]
+    stack = vsep [prettySrcLoc' loc | (_, loc) <- getCallStack callStack]
+
+-- | Adds a Doc to the log, which is only shown if the test fails
+logEntry :: HasCallStack => Doc AnsiStyle -> ImpTestM era ()
+logEntry = logWithCallStack ?callStack
```

Comment:
Let's keep the old version that uses `String`. It is not always convenient to work with `Doc`
```suggestion
logDocEntry :: HasCallStack => Doc AnsiStyle -> ImpTestM era ()
logDocEntry = logWithCallStack ?callStack

logEntry :: HasCallStack => String -> ImpTestM era ()
logEntry = logWithCallStack ?callStack . fromString
```